### DISCUSSION TITLE
[ 命名建議 ] LeetCode 977. Squares of a Sorted Array

### DISCUSSION THREAD

**Discussion by LPenny-github** [EngTW/English-for-Programmers#112] [Q&A] [answered]
LeetCode:
https://leetcode.com/problems/squares-of-a-sorted-array/

線上教材:
https://blog.techbridge.cc/2019/08/30/leetcode-pattern-two-pointer/

---
再麻煩 @twy30 給予命名建議 orz 感激不盡

```csharp
public int[] SortedSquares(int[] nums)
        {
            int[] input = nums;
            int[] output = new int[input.Length];
            int index = input.Length;
            int firstIndex = 0, secondIndex = index -1;
            do
            {
                int firstNumberSquare = input[firstIndex]*input[firstIndex];

                // 當兩個 index 相同，就在 output 加入該平方數後，送出答案
                if (firstIndex == secondIndex)
                {
                    output[--index] = firstNumberSquare;
                    return output;
                }

                int secondNumberSquare = input[secondIndex]*input[secondIndex];
                if (firstNumberSquare > secondNumberSquare)
                {
                    output[--index] = firstNumberSquare;
                    ++firstIndex;
                }
                else
                {
                    output[--index] = secondNumberSquare;
                    --secondIndex;
                }
            } while (firstIndex <= secondIndex);
            
           return new int[]{};            
        }
```

**Comment by twy30** [comment, marked as answer]
@LPenny-github 

你好 😊

---

```C#
            int index = input.Length;
```

* `index` 可以換成 `outputIndex`
  * `outputIndex` 初始化為 `output.Length - 1`
  * 在 `do-while` 迴圈中，從 `--index` 改成 `index--`
* 雖然 `input.Length` 與 `output.Length` 的值會是一樣的，但我會選使用 `output.Length` ，語意上比較一致 😊

---

```C#
            int firstIndex = 0, secondIndex = index -1;
```

```C#
                int firstNumberSquare = input[firstIndex]*input[firstIndex];
                int secondNumberSquare = input[secondIndex]*input[secondIndex];
```

在這個案例中，是從一個「一維陣列」的 頭/尾 、 始/終 兩個方向向中間夾擊，所以可以試試這樣做：

* `first` (第一), `second` (第二) 可以換成 `head` (頭), `tail` (尾)
  * 也可以換成 `start` (始), `end` (終)
  * 也可以與 `outputIndex` 呼應，寫成像這樣： `headInputIndex`, `tailInputIndex`
* 類似上一段提到的觀念，我會用 `input.Length - 1` 來初始化 `secondIndex` / `tailInputIndex`

可以參考看看 😊

**Reply by LPenny-github** [reply]
感謝 @twy30 大大 orz 

> 在 `do-while` 迴圈中，從 `--index` 改成 `index--`

為什麼要這樣做呢？

**Reply by twy30** [reply]
原本的寫法是這樣

```C#
            int index = input.Length;
```

在用到 `index` 時是 `--index` (先減 1 再取 `index` 的值)

---

如果依我的提議，把 `index` 的初始值設成 `output.Length - 1` ；因為已經先減了 1 ，所以用 `index--` (先取 `index` 的值，再將 `index` 減 1)

---

就「程式執行結果」來說，兩組方法的結果都會是正確的。

我個人的習慣是儘可能讓「索引變數」保持在 *合法範圍* 內，所以會傾向「把 `index` 的初始值設成 `output.Length - 1`」的寫法。

**Reply by LPenny-github** [reply]
@twy30  非常感謝 orz

**Reply by LPenny-github** [reply]
> `first` (第一), `second` (第二) 可以換成 `head` (頭), `tail` (尾)
也可以換成 start (始), end (終)

@twy30  附帶一提，我使用的教材是用 `left` (l)、`right` (r)😉
但因為我是個左右分不清的人，所以才一直使用 first 和 second 😅😆

總之，謝謝大大 😀

**Reply by twy30** [reply]
@LPenny-github `left` (左), `right` (右) 也是不錯的選擇 👍

很有意思的想法 😊
